sched/mqueue/msginternal.c

```diff
@@ -0,0 +1,137 @@
+ * Public Data
+ ****************************************************************************/
+
+struct list_node g_msgfreelist = LIST_INITIAL_VALUE(g_msgfreelist);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct list_node g_msgqactivelist =
+                        LIST_INITIAL_VALUE(g_msgqactivelist);
+static struct list_node g_msgqfreelist   =
+                        LIST_INITIAL_VALUE(g_msgqfreelist);
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxmsg_initialize
+ ****************************************************************************/
+
+void nxmsg_initialize(void)
+{
+  FAR struct msgbuf_s *msg;
+
+  msg = (FAR struct msgbuf_s *)kmm_malloc(sizeof(*msg) *
+                                          CONFIG_PREALLOC_MQ_MSGS);
+  if (msg)
+    {
+      int i;
+      for (i = 0; i < CONFIG_PREALLOC_MQ_MSGS; i++)
+        {
+          list_add_tail(&g_msgfreelist, &msg->node);
+          msg++;
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: nxmsg_alloc
+ ****************************************************************************/
+
+int nxmsg_alloc(FAR struct msgq_s **pmsgq)
+{
+  FAR struct msgq_s *msgq;
+
+  /* Allocate memory for the new message queue. */
+
+  msgq = list_remove_head_type(&g_msgqfreelist, struct msgq_s, node);
+  if (msgq == NULL)
+    {
+      msgq = (FAR struct msgq_s *)kmm_zalloc(sizeof(struct msgq_s));
+    }
+
+  if (msgq == NULL)
+    {
+      return -ENOSPC;
+    }
+
+  /* Initialize the new named message queue */
+
+  list_add_tail(&g_msgqactivelist, &msgq->node);
+  list_initialize(&msgq->msglist);
+
+  msgq->maxmsgs    = MSG_MAX_MSGS;
+  msgq->maxmsgsize = MSG_MAX_BYTES;
+
+  *pmsgq = msgq;
+  return OK;
+}
+
+/****************************************************************************
+ * Name: nxmsg_free
+ ****************************************************************************/
+
+void nxmsg_free(FAR struct msgq_s *msgq)
+{
+  FAR struct msgq_s *entry;
+  FAR struct msgq_s *tmp;
+
+  list_for_every_entry_safe(&msgq->msglist, entry, tmp, struct msgq_s, node)
+    {
+      list_delete(&entry->node);
+      list_add_tail(&g_msgfreelist, &entry->node);
+    }
+
+  list_delete(&msgq->node);
+  list_add_tail(&g_msgqfreelist, &msgq->node);
+}
+
+/****************************************************************************
+ * Name: nxmsg_lookup
+ ****************************************************************************/
+
+FAR struct msgq_s *nxmsg_lookup(key_t key)
+{
+  FAR struct msgq_s *msgq;
+
+  list_for_every_entry(&g_msgqactivelist, msgq, struct msgq_s, node)
```

Comment:
can we change g_msgqactivelist to an array and identify the msq by index to improve the speed?

Reply:
Done